src/main/java/org/opensearch/knn/index/mapper/ANNConfig.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.mapper;
+
+import lombok.Getter;
+import org.opensearch.knn.index.engine.KNNMethodContext;
+
+import java.util.Optional;
+
+/**
+ * Class holds information about how the ANN indices are created. The design of this class ensures that we do not
+ * accidentally configure an index that has multiple ways it can be created. This class is immutable.
+ */
+public final class ANNConfig {
+
+    @Getter
+    private final ANNConfigType annConfigType;
```

Comment:
Rather than a type here, is it worth simplifying it to have a `boolean requiresTraining()`

Reply:
There is a third case where flatmapper when knn=false where it is not model based or method based. Working on cleaning this up a little bit.